### DISCUSSION TITLE
Intellisense documentation and autosuggestions lost when decorating class

### DISCUSSION THREAD

**Discussion by tarheels100** [microsoft/pyright#4488] [Q&A] [answered]
Hello,

I have a simple class decorator that provides a default `__repr__` implementation:
```python
class AutoRepr:
    """Decorator that generates provides a general-purpose repr implementation.
    """
    @staticmethod
    def repr(obj) -> str:
        """General-purpose repr implementation.

        Args:
            obj (Any): the object to generate representation of

        Returns:
            str: object representation as string
        """
        items = []
        for prop, value in obj.__dict__.items():
            try:
                item = "%s = %r" % (prop, value)
                assert len(item) < 10000
            except:
                item = "%s: <%s>" % (prop, value.__class__.__name__)
            items.append(item)

        return "%s(%s)" % (obj.__class__.__name__, ', '.join(items))

    def __init__(self, cls):
        cls.__repr__ = AutoRepr.repr
        self.cls = cls

    def __call__(self, *args, **kwargs):
        return self.cls(*args, **kwargs)
```
When I apply this decoration to a class, I lose all intellisense documentation and autosuggestions in VSCode with pylance.  Hovering over an object instance, I get something like this: `(variable) foo: Any`.  Hovering over the class name when instantiating an object, I get `(class) bar` without any additional doc string info.  Is there a way to fix this?

PS: I initially thought that [this](https://github.com/microsoft/pyright/discussions/4426) might work, but I'm not sure that method applies in my case.

**Comment by erictraut** [comment, marked as answer]
Here's how to annotate such a decorator:

```python
from typing import Generic, ParamSpec, TypeVar

P = ParamSpec("P")
R = TypeVar("R")


class AutoRepr(Generic[P, R]):
    @staticmethod
    def repr(obj) -> str:
        ...

    def __init__(self, cls: Callable[P, R]):
        cls.__repr__ = AutoRepr.repr
        self.cls = cls

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
        return self.cls(*args, **kwargs)
```

**Reply by tarheels100** [reply]
@erictraut Wow, thanks for the fast response!  I implemented your suggestion and it seems to fix autosuggestions for object attributes and methods, but for some reason I'm still not getting the doc string to show up like it does when the decorator is removed.  
With decorator:
![with decorator](https://user-images.githubusercontent.com/29213686/213283554-8d763c25-aacd-4599-908c-8a2ae57c24f4.png)
Without decorator:
![no decorator](https://user-images.githubusercontent.com/29213686/213283579-109a6de9-c8ad-45d8-ab8f-9ab6adfb1bd2.png)

Maybe I've got some other issue going on as well.

**Reply by erictraut** [reply]
Pyright doesn't currently capture the class or constructor doc string in this case, but for consistency it should. I've just updated the pyright logic to handle this. This will be included in the next release.
